Add slider velocity slider to slider toolbox

chris-ehmann opened the pull request:
(a.k.a., how many times can I use the word "slider" in one PR?)

Closes #29309
Closes https://github.com/ppy/osu/pull/33707

This PR adds a slider to control slider velocity to the freehand slider toolbox group. Even though this toolbox group currently contains controls solely for freehand drawing of sliders, I think adding the control for slider velocity in here makes the most sense. It would feel kind of weird to have one group labelled "Slider", and then another section solely for slider velocity.

I considered adding a toggle to couple the value from this slider to distance snapping (like mentioned [here](https://github.com/ppy/osu/issues/29309#issuecomment-2747068625)) in this PR, but a) I would like to get confirmation on whether such a toggle is necessary before doing so and b) I would rather split that into a separate PR, anyways.

## Quick Demo

https://github.com/user-attachments/assets/2dacc4d5-c5c0-4594-bcaf-4195df547739

chris-ehmann committed:
Add slider for Slider Velocity

chris-ehmann committed:
remove no longer needed slider velocity code

chris-ehmann committed:
remove unused variable

minetoblend reviewed (commented):
This essentially removes the behavior of inheriting slider velocity from the previous object, which brings up some UX concerns about having to change the velocity manually all the time when seeking back and forth between high/low SV sections.
I think it's worth considering having some kinda of mechanism to update the slider based on the surrounding HitObjects.

chris-ehmann commented on osu.Game.Rulesets.Osu/Edit/FreehandSliderToolboxGroup.cs:
Agreed on this, maybe just shortening the name to `SliderToolboxGroup` would be good?

chris-ehmann commented on osu.Game.Rulesets.Osu/Edit/FreehandSliderToolboxGroup.cs:
```diff
@@ -85,6 +97,14 @@ protected override void LoadComplete()
+                sliderVelocitySlider.ContractedLabelText = $"S. V.: {velocity.NewValue:N2}";
+                sliderVelocitySlider.ExpandedLabelText = $"Slider Velocity: {velocity.NewValue:N2}";
+
+                SliderVelocity.Value = velocity.NewValue;
```
oops, removed 😅

chris-ehmann committed:
Rename FreehandSliderToolboxGroup to SliderToolboxGroup, remove line assigning bindable to itself

chris-ehmann committed:
code quality fix

chris-ehmann committed:
multiple improvements to slider

chris-ehmann committed:
Add button to set s.v. value to previous slideron timeline's value

minetoblend commented on osu.Game.Rulesets.Osu/Edit/SliderToolboxGroup.cs:
```diff
@@ -80,6 +90,16 @@ private void load()
+                {
+                    Action = () =>
+                    {
+                        double? previousSliderVelocity = getPreviousSliderVelocityMultiplier();
```
I feel like it might be a good idea to store the previous slider velocity in a field/bindable. Input/clock update happen before this component's update loop runs so the value could be no longer available on button press.

minetoblend commented on osu.Game.Rulesets.Osu/Edit/SliderToolboxGroup.cs:
```diff
@@ -148,5 +168,31 @@ protected override void LoadComplete()
+        {
+            return (editorBeatmap
+                    .HitObjects
+                    .LastOrDefault(h => h is Slider && h.StartTime < editorClock.CurrentTime) as Slider)?.SliderVelocityMultiplier;
```
Given that this slider is used for placing new sliders only it mightt make sense to do a `<=` check here for the startTime. If we place a slider on top of another slider I think the most likely outcome we want is to inherit *that* slider's velocity

chris-ehmann committed:
store previous slider velocity in private field + update to check if previous slider start time equal to current time

chris-ehmann committed:
Code quality fixes

chris-ehmann committed:
Actually fix quality this time (oops)

chris-ehmann committed:
Add back line applying defaults to slider + rewrite comment

I had initially thought this line was irrelevant given the initial comment,
but it is still necessary to apply defaults to sliders to avoid showing an
incorrect slider length in the timeline.

peppy reviewed (commented):
Not sure on behaviour change. Sure, it matches stable, but I don't know if it's the most intuitive answer here.

bdach commented on osu.Game.Rulesets.Osu/Edit/SliderToolboxGroup.cs:
```diff
@@ -57,15 +70,36 @@ public FreehandSliderToolboxGroup()
         {
             Children = new Drawable[]
             {
+                sliderVelocitySlider = new ExpandableSlider<double, SizeSlider<double>>
```
Possibly out of scope, but really think this should come with a textbox, too, if easily possible. Just slider is a bit annoying here.

chris-ehmann replied:
Good call. I'll try working on some sort of expandable slider w/ textbox - I don't think it should be too bad to add.

bdach reviewed (commented):
Thanks for trying this one. Maybe it'll get one guy to get off my case and stop complaining to me that I'm "forcing them to practice math" or whatever.

bdach commented on osu.Game.Rulesets.Osu/Edit/SliderToolboxGroup.cs:
```diff
@@ -128,5 +168,26 @@ protected override void LoadComplete()
+                                      .HitObjects
+                                      .LastOrDefault(h => h is Slider && h.StartTime <= editorClock.CurrentTime) as Slider)?.SliderVelocityMultiplier;
+
+            if (previousSliderVelocity != null)
```
> Not sure on behaviour change. Sure, it matches stable

Pretty sure it does not, stable does factor in slider velocity by default doesn't it?

chris-ehmann replied:
My initial idea was to have the slider be updated with the value of the previous slider's velocity as the user navigates through the map to essentially mimic the current behaviour (see below video), but I was concerned this might be annoying in certain scenarios. For example: the user interacts with the slider to set a value they want to use, but then they seek to another part of the map/accidentally scroll, causing the value they've set to be overwritten. I would like to use this approach, but I'm not entirely sure about handling the aforementioned issue (if it's even that big of a concern).

https://github.com/user-attachments/assets/80ca78a8-66ec-4ebb-bae3-d98beb74315c

FWIW, I also thought about having the slider bar value not be connected to the previous slider, but I think that gets into weird territory where it's very unclear what value should be used (e.g., a user never touches the slider bar -> previous slider value is used, but after they interact with the slider bar, when does the slider bar value **stop** being used?)
